README.md
# IPFS-WebDAV

IPFS-WebDAV is a daemon that uses [IPFS] HTTP RPC and exposes [WebDAV] API. WebDAV, in turn, can be used
to mount the IPFS file system in most operating systems.

```
Linux   -------------->   +-------------+           IPFS HTTP RPC           +------+
Mac OS  -----WebDAV-----> | IPFS-WebDAV | --------------------------------> | IPFS |
Windows -------------->   +-------------+   (e.g. http://localhost:5001)    +------+
```

Currently, IPFS-WebDAV exposes `/ipfs` and `/ipns` namespaces in read-only mode, and `/mfs` namespace
(referring to [MFS]) in read-write mode. The read-write access for `/ipns` is planned.

[IPFS]: https://ipfs.io
[WebDAV]: https://en.wikipedia.org/wiki/WebDAV
[MFS]: https://docs.ipfs.io/concepts/file-systems/

## Advantages over IPFS mount

It's no secret that IPFS [can mount] `/ipfs` and `/ipns` out-of-the-box.
However, using `ipfs-webdav` has several advantages:
- WebDAV is supported natively in most operating systems
- You can run single `IPFS` + `ipfs-webdav` node and connect to it from multiple hosts using WebDAV.
  This is useful in LAN or in cloud
- `ipfs-webdav` provide read-write access to [MFS] 
- You can use `ipfs name publish` along with accessing `/ipns` via WebDAV. 

[can mount]: https://github.com/ipfs/go-ipfs/blob/a72753bade90c4a48c29aba6c0dc81c44785e9d2/docs/fuse.md
[MFS]: https://docs.ipfs.io/concepts/file-systems/

## Installation

### Docker

[![Docker Image Version (latest semver)](https://img.shields.io/docker/v/vlad20012/ipfs-webdav?arch=amd64&color=blue&label=ipfs-webdav%20docker%20image&sort=date)](https://hub.docker.com/r/vlad20012/ipfs-webdav)

Follow the guide on the [official docker image page](https://hub.docker.com/r/vlad20012/ipfs-webdav).

### Installation from binary distribution

You can find `ipfs-webdav` binaries on the [releases page](https://github.com/vlad20012/ipfs-webdav/releases).
Then just run it (`./ipfs-webdav_linux-x86-64`)

### Installation From Source

Run `cargo build --release`, then find the binary in `target/release/ipfs-webdav`

## Configuration

Currently, `ipfs-webdav` can be configured only using environment variables.
- `IPFS_WEBDAV_API_ENDPOINT_URL` - specifies URL of an IPFS RPC. For example, `http://localhost:5001`.
  If not specified, `ipfs-webdav` tries to read the URL from `~/.ipfs/api` file and falls back to
  `http://localhost:5001` if the file is not found.
- `IPFS_WEBDAV_LISTEN` - Specifies listen WebDAV address in `host:port` format. 
  For example, `localhost:4918` or `0.0.0.0:4918`. Default value is `127.0.0.1:4918`
- `IPFS_WEBDAV_LOG` - specifies `ipfs-webdav` log level. Possible values: `error`, `warn`, `info`, `debug`, `trace`.
  Default value is `info`.

Example: `IPFS_WEBDAV_API_ENDPOINT_URL="http://localhost:5001" IPFS_WEBDAV_LISTEN="0.0.0.0:4918" ./ipfs-webdav`

## Mounting WebDAV filesystem

### Linux

The simplest way to connect WebDAV on Linux with GNOME is connecting from `Nautilus` (aka `Files`).
Open Nautilus, select "+ Other Locations" on the left sidebar then type your server address in the
"Connect to Server" field. The address is usually `dav://localhost:4918/` or whatever you specified
using `IPFS_WEBDAV_LISTEN` environment variable. Then press `Connect` button.

For long-living mounts and for other Linux environments I suggest using `davfs2`:
```bash
mount -t davfs http://127.0.0.1:4918 /mnt
```
It asks to enter username and then password, just press `<enter>` instead.

### MacOS

On macOS, you can simply connect WebDAV drive from `Finder`. Follow the 
[official guide](https://support.apple.com/guide/mac-help/connect-disconnect-a-webdav-server-mac-mchlp1546/mac).
Use `http://localhost:4918/` (or whatever you specified using `IPFS_WEBDAV_LISTEN` environment variable) as
"Server Address".

### Windows

On Windows, you can simply connect WebDAV drive from `File Explorer` using `Map network drive...`
option. Use `http://localhost:4918/` (or whatever you specified using `IPFS_WEBDAV_LISTEN` environment variable) as 
"Internet or network address".

## License

The IPFS-WebDAV project is dual-licensed under Apache 2.0 and MIT terms:

- Apache License, Version 2.0, ([LICENSE-APACHE](https://github.com/vlad20012/ipfs-webdav/blob/master/LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
- MIT license ([LICENSE-MIT](https://github.com/vlad20012/ipfs-webdav/blob/master/LICENSE-MIT) or http://opensource.org/licenses/MIT)


CHANGELOG.md
# 0.1.1 (27.04.2023)

- Fix issue [`#2`](https://github.com/vlad20012/ipfs-webdav/issues/2) - after uploading, 
  files are defective and larger
- Update dependencies

# 0.1.0 (13.05.2022)

- Initial release


src/main.rs
mod ipfs_api_ext;
mod ipfs_fs;
mod metadata;
mod mfs;
mod read;
mod request_log;
mod rootfs;
mod write;

use crate::ipfs_fs::IpfsFs;
use crate::metadata::StaticMetadata;
use crate::mfs::MfsFs;
use crate::request_log::RequestLog;
use crate::rootfs::RootFs;
use futures::TryStreamExt;
use hyper::Request;
use ipfs_api_backend_hyper::{Error, IpfsApi, IpfsClient, TryFromUri};
use std::convert::Infallible;
use std::error::Error as _;
use std::net::SocketAddr;
use std::str::FromStr;
use unixfs_v1::dagpb::node_data;
use unixfs_v1::UnixFs;
use webdav_handler::davpath::DavPath;
use webdav_handler::fs::FsError;
use webdav_handler::memls::MemLs;
use webdav_handler::DavHandler;

#[tokio::main]
async fn main() {
    env_logger::Builder::from_env(
        env_logger::Env::default()
            .filter_or("IPFS_WEBDAV_LOG", "info")
            .write_style("IPFS_WEBDAV_LOG_STYLE"),
    )
    .init();

    let ipfs = std::env::var("IPFS_WEBDAV_API_ENDPOINT_URL")
        .ok()
        .map(|addr| {
            IpfsClient::from_str(&addr).unwrap_or_else(|e| {
                log::error!(
                    "Failed to parse IPFS API endpoint URL `{}` specified by \
                `IPFS_WEBDAV_API_ENDPOINT_URL` environment variable. It must be \
                in URL form. For example, `http://localhost:5001`. Error details: {}",
                    addr,
                    e
                );
                std::process::exit(101)
            })
        })
        .unwrap_or_else(|| IpfsClient::default());

    let addr: SocketAddr = std::env::var("IPFS_WEBDAV_LISTEN")
        .ok()
        .map(|addr| {
            SocketAddr::from_str(&addr).unwrap_or_else(|_| {
                log::error!(
                    "Failed to parse listen address `{}` specified by \
                `IPFS_WEBDAV_LISTEN` environment variable. It must be in `host:port` \
                form. For example, `localhost:4918` or `0.0.0.0:4918`",
                    addr
                );
                std::process::exit(101)
            })
        })
        .unwrap_or_else(|| ([127, 0, 0, 1], 4918).into());

    let dav_server = DavHandler::builder()
        .autoindex(true)
        .filesystem(Box::new(RootFs {
            mfs: MfsFs { ipfs: ipfs.clone() },
            ipfs: IpfsFs {
                ipfs: ipfs.clone(),
                ty: IpfsOrIpns::Ipfs,
            },
            ipns: IpfsFs {
                ipfs,
                ty: IpfsOrIpns::Ipns,
            },
        }))
        .locksystem(MemLs::new())
        .build_handler();

    let make_service = hyper::service::make_service_fn(move |_| {
        let dav_server = dav_server.clone();
        async move {
            let func = move |req: Request<_>| {
                let log = RequestLog::on_request(&req);
                let dav_server = dav_server.clone();
                async move {
                    let resp = dav_server.handle(req).await;
                    log.on_response(&resp);
                    Ok::<_, Infallible>(resp)
                }
            };
            Ok::<_, Infallible>(hyper::service::service_fn(func))
        }
    });

    let server = hyper::Server::try_bind(&addr).unwrap_or_else(|e| {
        match e.source() {
            None => log::error!("Error binding to {}: {}", addr, e),
            Some(s) => log::error!("Error binding to {}: {}", addr, s),
        }
        std::process::exit(101)
    });

    log::info!("Starting WebDAV server at http://{}", addr);

    let _ = server
        .serve(make_service)
        .await
        .map_err(|e| log::error!("server error: {}", e));
}

#[derive(Clone)]
pub enum IpfsOrIpns {
    Ipns,
    Ipfs,
}

async fn stat_metadata(ipfs: &IpfsClient, ipfs_path: &str) -> Result<StaticMetadata, FsError> {
    let response = ipfs
        .block_get(ipfs_path)
        .map_ok(|chunk| chunk.to_vec())
        .try_concat()
        .await;
    let block_bytes = response.map_err(handle_error)?;
    let unixfs_data_result =
        node_data(&block_bytes).and_then(|block_data| UnixFs::try_from(block_data));
    let data = match unixfs_data_result {
        Ok(data) => data,
        Err(_) => {
            return Ok(StaticMetadata {
                len: block_bytes.len() as u64,
                is_dir: false,
                is_unixfs: false,
            })
        }
    };
    Ok(StaticMetadata::from_unixfs_data(&data))
}

fn map_path(path: &DavPath) -> Result<&str, FsError> {
    std::str::from_utf8(path.as_bytes()).map_err(|_| FsError::GeneralFailure)
}

fn handle_error(e: Error) -> FsError {
    match &e {
        Error::Api(e) => {
            if e.code == 0
                && (e.message == "file does not exist" || e.message.starts_with("no link named "))
            {
                return FsError::NotFound;
            }
        }
        _ => {}
    };

    log::error!("Got an error from IPFS API: {}", e);

    return FsError::GeneralFailure;
}
